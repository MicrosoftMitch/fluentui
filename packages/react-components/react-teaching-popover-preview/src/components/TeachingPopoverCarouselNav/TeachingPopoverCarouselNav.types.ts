import { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type TeachingPopoverCarouselNavSlots = {
  /**
   * The element wrapping the carousel pagination. By default this is a div,
   * it may contain icons or text depending on TeachingPopoverCarouselNavStyle
   */
  root: NonNullable<Slot<'div'>>;
};

export type TeachingPopoverCarouselNavState = ComponentState<TeachingPopoverCarouselNavSlots> & {};

export type TeachingPopoverCarouselNavProps = ComponentProps<Partial<TeachingPopoverCarouselNavSlots>>;
