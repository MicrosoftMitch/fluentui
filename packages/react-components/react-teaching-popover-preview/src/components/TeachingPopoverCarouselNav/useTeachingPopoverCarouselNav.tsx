import * as React from 'react';
import { getIntrinsicElementProps, slot } from '@fluentui/react-utilities';
import type {
  TeachingPopoverCarouselNavProps,
  TeachingPopoverCarouselNavState,
} from './TeachingPopoverCarouselNav.types';

import { useArrowNavigationGroup } from '@fluentui/react-tabster';
import { TeachingPopoverCarouselNavButton } from '../TeachingPopoverCarouselNavButton/index';
import { useCarouselContext_unstable } from '../TeachingPopoverCarousel/Carousel/CarouselContext';
import { useSyncExternalStore } from 'use-sync-external-store/shim';

/**
 * Returns the props and state required to render the component
 * @param props - TeachingPopoverCarouselNav properties
 * @param ref - reference to root HTMLElement of TeachingPopoverCarouselNav
 */
export const useTeachingPopoverCarouselNav_unstable = (
  props: TeachingPopoverCarouselNavProps,
  ref: React.Ref<HTMLDivElement>,
): TeachingPopoverCarouselNavState => {
  const focusableGroupAttr = useArrowNavigationGroup({
    circular: false,
    axis: 'horizontal',
    memorizeCurrent: false,
    // eslint-disable-next-line @typescript-eslint/naming-convention
    unstable_hasDefault: true,
  });

  const store = useCarouselContext_unstable(c => c.store);
  const values = useSyncExternalStore(store.subscribe, () => store.getSnapshot());

  // Generate the child TeachingPopoverCarouselNavButton and memoize them to prevent unnecessary re-rendering
  const rootChildren = React.useMemo(() => {
    return Array.from(Array(values.length), (_, i) => <TeachingPopoverCarouselNavButton key={i} value={values[i]} />);
  }, [values]);

  return {
    components: {
      root: 'div',
    },
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref,
        role: 'tablist',
        tabIndex: 0,
        ...props,
        children: rootChildren,
        ...focusableGroupAttr,
      }),
      { elementType: 'div' },
    ),
  };
};
