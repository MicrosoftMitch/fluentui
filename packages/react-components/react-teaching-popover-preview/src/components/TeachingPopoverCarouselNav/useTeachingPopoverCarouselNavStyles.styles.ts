import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type {
  TeachingPopoverCarouselNavSlots,
  TeachingPopoverCarouselNavState,
} from './TeachingPopoverCarouselNav.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';
import { tokens } from '@fluentui/react-theme';

export const teachingPopoverCarouselNavClassNames: SlotClassNames<TeachingPopoverCarouselNavSlots> = {
  root: 'fui-TeachingPopoverCarouselNav',
};

const useStyles = makeStyles({
  root: {
    display: 'flex',
    flexDirection: 'row',
    columnGap: '4px',
    alignItems: 'center',
    justifyContent: 'center',
    ...createCustomFocusIndicatorStyle({
      ...shorthands.outline('2px', 'solid', tokens.colorStrokeFocus2),
      ...shorthands.borderRadius('4px'),
      ...shorthands.borderColor('transparent'),
    }),
  },
});

/** Applies style classnames to slots */
export const useTeachingPopoverCarouselNavStyles_unstable = (state: TeachingPopoverCarouselNavState) => {
  const styles = useStyles();

  state.root.className = mergeClasses(teachingPopoverCarouselNavClassNames.root, styles.root, state.root.className);

  return state;
};
