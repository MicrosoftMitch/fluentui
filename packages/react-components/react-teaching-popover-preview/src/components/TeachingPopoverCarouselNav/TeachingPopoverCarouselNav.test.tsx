import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { TeachingPopoverCarouselNav } from './TeachingPopoverCarouselNav';

describe('TeachingPopoverCarouselNav', () => {
  isConformant({
    Component: TeachingPopoverCarouselNav,
    displayName: 'TeachingPopoverCarouselNav',
  });

  it('renders a default state', () => {
    const result = render(<TeachingPopoverCarouselNav>Default TeachingPopoverCarouselNav</TeachingPopoverCarouselNav>);
    expect(result.container).toMatchSnapshot();
  });
});
