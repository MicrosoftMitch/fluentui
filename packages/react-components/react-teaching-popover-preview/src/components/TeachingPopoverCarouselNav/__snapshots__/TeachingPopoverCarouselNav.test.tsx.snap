// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeachingPopoverCarouselNav renders a default state 1`] = `
<div>
  <div
    class="fui-TeachingPopoverCarouselNav"
    data-tabster="{\\"groupper\\":{\\"tabbability\\":1},\\"focusable\\":{}}"
    role="list"
    tabindex="0"
  >
    <div
      class="fui-TeachingPopoverCarouselNavIcon"
    >
      <button
        class="fui-TeachingPopoverCarouselNavIcon__navButton"
        type="button"
      />
    </div>
    <div
      class="fui-TeachingPopoverCarouselNavIcon"
    >
      <button
        class="fui-TeachingPopoverCarouselNavIcon__navButton"
        type="button"
      />
    </div>
  </div>
</div>
`;
